Implement `getPapersToReviewForProject` call

domienderle opened the pull request:
Closes #99 

## What I have made

<!-- write down what changes have been made, what was removed/added/changed -->
<!-- e.g. I added a service, which does x -->

**Note:** Only the last four commits are relevant for this PR. Since this PR is based on many functionality, which is implemented in #181  I rebased this branch on #257, which will be merged soon.

**Note:** This PR can already be reviewed even though the formatter fails. Depending on the review I will clean up the code for the formatter.

**Note:** For the reviewer: Please have a look into the service implementation of the `getPapersToReviewForProject` method. I am not completely happy with this implementation, since we would have to iterate over the project papers two times, but this way we can avoid a lot of duplicated code and I did not find a good way to implement this with getting the best of both worlds. Maybe you can have a look and let me know if you have any good ideas.

I have implemented the `getPapersToReviewForProject` method by refactoring the `getAllProjectPapersForProject` method and adding a bit of functinality.

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does
not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] (for reviewer) I have checked the implementation against the requirements

domienderle committed:
feat: add service layer for getPapersToReviewForProject call

domienderle committed:
fix: change mapping of from papers to review to project papers to review

domienderle committed:
feat: add grpc layer for getPapersToReviewForProject call

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/model/dto/ProjectPaperWithPaper.kt:
One general hint that does not directly is related to your changes: if you import ProjectOuterClass.Project.Paper directly as GrpcPaper then the code gets a bit shorter and cleaner in my opinion and you sometimes use `...` and sometimes [...] to reference to another type or so. Maybe this can be unified.

Merseleo replied:
And you can shorten the projectPaper.toGrpcProjectPaper(...) function by only providing the arguments.

domienderle replied:
You are right with all of that. I hope I got everything.

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -29,11 +30,12 @@ import se.uulm.snowballr.backend.repository.association.IReviewTableRepo
 import snowballr.ProjectOuterClass.Project as GrpcProject
 
-@Suppress("ComplexInterface")
+@Suppress("ComplexInterface", "TooManyFunctions")
```
Actually this should be an indicator for us, that there are probably functions in the _project_ service that actually doesn't belong here. As I already wrote in Mattermost, I am not happy with how we split the services and group the functions. Actually this service should be for all project related tasks and in my opinion we should have either a separate paper / project paper service or a project paper / review service. Maybe we could discuss this again in group, e.g., tomorrow or so.

domienderle replied:
Yes I agree, the separation of the service is not really good at the moment. We have to discuss this.

Merseleo replied:
Ok 👍

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -278,7 +285,14 @@ class ProjectService(
+        val paperReviewsMap: Map<ProjectPaper, List<ReviewOuterClass.Review>>,
+    )
+
+    private suspend fun getProjectPapers(request: Base.Id): ProjectPaperData {
```
At this point, I have a few requests: 1) if we do not split the project paper calls from the project calls, then we should stay consistent and move this data class and private function up, i.e., before all service call implementations. 2) in my opinion projectPaperReviewsMap would be more correct to indicate that you map project papers and not normal papers to their reviews and 3) in my opinion the wording `getProjectPapers` is a bit misleading as (and the return type indicates) you get only some data of the project papers but not the project papers itself, right? So maybe something like `getProjectPaperData` would be more precise?

domienderle replied:
I tried to include everything here. I have changed the `getProjectPapers` function to get an optional predicate function, as you suggested (thanks for the good suggestion). Thus, the function now really returns the `GrpcProjectPapers`, which is why the naming now makes sense imo. I also moved it up.

Merseleo replied:
Yes, in my opinion this was a very good update 👍 and I agree, now the name is completely right. Good gob!

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -307,14 +321,35 @@ class ProjectService(
+
+            !isAlreadyReviewed &&
+                it.projectPaper.decision != ProjectOuterClass.PaperDecision.PAPER_DECISION_ACCEPTED &&
+                it.projectPaper.decision != ProjectOuterClass.PaperDecision.PAPER_DECISION_DECLINED
```
I would invert this condition, so check whether it is unreviewed or in review, because theoretically a paper decision could also be `PAPER_DECISION_UNSPECIFIED` and then it would be included. So instead of "checking a blacklist" I would suggest to "check the white list"?

domienderle replied:
Good point.

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -278,7 +285,14 @@ class ProjectService(
+    private suspend fun getProjectPapers(request: Base.Id): ProjectPaperData {
         val currentUser = userRepo.getUserById(GrpcContext.getUserIdFromContext())
         val projectId = parseUUID(request.id, EntityType.PROJECT)
         repo.getProjectById(projectId)
```
In my opinion it would be more correct, if you add a function `doesProjectExist` or at least add a comment explaining why you retrieve a project without using it.

domienderle replied:
I added a seperate `deosProjectExist` function and since we need this type of functions more often in the future I created a generic function for all tables. What do you think about the implementation?

Merseleo replied:
Very good idea, I like it a lot.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/GetPapersToReviewForProjectTest.kt:
```diff
@@ -0,0 +1,293 @@
+import java.util.stream.Stream
+import kotlin.reflect.KFunction
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
```
Why you set the lifecycle to `PER_CLASS`?

domienderle replied:
Same as in #267. I have to, because otherwise I cannot paramerterize the tests.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/GetPapersToReviewForProjectTest.kt:
```diff
@@ -0,0 +1,293 @@
+        coEvery { reviewRepoMock.getAllReviewsForProjectPaper(projectPaper.id) } returns listOf(review)
+        coEvery {
+            reviewHasCriterionRepoMock.getSelectedCriteriaIdsForReviewById(review.id)
+        } returns listOf(UUID.randomUUID())
```
Isnt it possible and useful to add a parameter to your happyPath function that changes to role of the user in the happy path and reuse the code? Because it is very very similar.

domienderle replied:
You're right. I have added the parameter.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/GetPapersToReviewForProjectTest.kt:
```diff
@@ -0,0 +1,293 @@
+        }
+
+    @Test
+    fun `When the project papers to review are requested, then only the not already decided papers are returned`() =
```
"then only the undecided papers" ?

domienderle replied:
Thanks, that was the formulation my bad english was looking for.

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/GetPapersToReviewForProjectTest.kt:
```diff
@@ -0,0 +1,293 @@
+        }
+
+    @Test
+    fun `When the project papers to review are requested, then only the papers that were not already reviewed by the current user are returned`() =
```
maybe rewrite it to something like "then only undecided papers that were not already reviewed ..."?

Merseleo commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/GetPapersToReviewForProjectTest.kt:
```diff
@@ -0,0 +1,293 @@
+    @Test
+    fun `When the project papers to review are requested, then only the papers that were not already reviewed by the current user are returned`() =
+        runTest {
+            val currentUser = DataBuilder.createExampleUser(role = UserOuterClass.UserRole.USER_ROLE_ADMIN)
```
you can drop the specification of the role, right?

domienderle replied:
No I can't, because otherwise an exception is thrown, that the user is not a project member.

Merseleo replied:
Ah right and probably it is easier to set the user as admin instead of making it a project member.

Merseleo reviewed (changes requested):
Of course I do not know what will be changed in the previous commits, but in general the feature seems to work. However, (but this can be also personal preference), the internals of the feature have potential for improvement. 

In general, I very like the idea that you reuse the code and avoid duplicate code. 
And in my opinion we should prefer readable and maintainable code instead of start optimizing now. I dont think that it is a problem that we iterate two times over the project papers. But if you are not happy, then I can only suggest, that your `getProjectPapers` function also takes an optional predicate function and if it is given it is applied to directly filter the papers. This way you would only iterate once. 

And one last thing: please do not forget to drop the last commit after the review.

domienderle committed:
feat: refactor getPapers function

domienderle committed:
feat: add doesProjectExits function and adjust tests

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/model/dto/ProjectPaperWithPaper.kt:
```diff
@@ -1,37 +1,37 @@
- * @param reviews A list of reviews represented as `ReviewOuterClass.Review`, associated with the paper.
- * @return A `ProjectOuterClass.Project.Paper` object constructed with data from the `ProjectPaperWithPaper` instance and the provided authors, backward references, and reviews.
+ * @param authors A list of authors represented as [GrpcAuthor], associated with the [Paper].
+ * @param backwardReferencedIds A list of strings representing the IDs of [Paper]s referenced by the current [Paper].
```
So if I understood our discussion in #267 right, then this should be "the IDs of [Paper]s referencing the current [Paper], right? Or am I completely confused now?

domienderle replied:
No this is correct. The backward references are the papers that are referenced by the current paper. The forward references are the Papers that cite the current paper.

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -75,13 +109,93 @@ interface IProjectService {
+    private val reviewHasCriterionTableRepo: IReviewHasCriterionTableRepo,
 ) : IProjectService {
+    /**
+     * Retrieves a list of [ProjectPaper] associated with a specified [Project] ID. This method
```
```suggestion
     * Retrieves a list of [ProjectPaper] associated with a specified project id. This method
```

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -75,13 +109,93 @@ interface IProjectService {
 ) : IProjectService {
+    /**
+     * Retrieves a list of [ProjectPaper] associated with a specified [Project] ID. This method
+     * also ensures access control for the current user. Optionally, a predicate can be provided to filter
```
```suggestion
     * also ensures access control for the current user. Optionally, a predicate function can be provided to filter
```

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -75,13 +109,93 @@ interface IProjectService {
+     * @param request The request containing the ID of the [Project] for which [ProjectPaper]s are to be retrieved.
+     * @param predicate An optional lambda function that takes a [ProjectPaperWithPaper], a map of [GrpcReview], and a user ID.
+     * This function should return a boolean value to filter the [ProjectPaperWithPaper]s. If null, no filtering is applied.
+     * @return A list of [GrpcProjectPaper] as a [GrpcProjectPaper.List] including associated metadata such as authors,
```
```suggestion
     * @return A list of [GrpcProjectPaper] including associated metadata such as authors,
```

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -205,4 +319,54 @@ class ProjectService(
+            { projectPaper, projectPaperReviewsMap, currentUserId ->
+                val isAlreadyReviewedByCurrentUser = projectPaperReviewsMap[projectPaper.projectPaper]
+                    ?.any { review -> review.userId == currentUserId } == true
+                val isAlreadyDecided =
```
```suggestion
                val isStillUndecided =
```
and when you use it as well. Because now you inverted the check

domienderle replied:
True. Thanks.

Merseleo commented on src/main/kotlin/se/uulm/snowballr/backend/repository/ProjectTableRepo.kt:
```diff
@@ -34,6 +34,11 @@ interface IProjectTableRepo {
     suspend fun getProjectById(id: UUID): Project
 
+    /**
+     * @return whether the project with the passed [id] exists.
```
```suggestion
     * Checks whether the project with the passed [id] exists.
```

Merseleo reviewed (changes requested):
Ok all in all, I think we are nearly finished. But I would suggest, that you finish the other PRs first, rebase and then I will have a final look and test.

domienderle committed:
chore: include last review changes

domienderle commented:
All right, I rebased the branch on develop as the other PR is done. I also dropped the test setup commit, so we should be ready to merge here, if you don't have anything else.

Merseleo reviewed (dismissed):
Ok, as the implementation of the `getBackwardReferences...` call in the repository layer is fixed in #267, everything should be fine. I double checked everything, but 1) everything works and 2) looks good. 
If you want, you add some missing tests (I think regarding the doesEntityExist or doesProjectExist thing, check the coverage report), but these tests are not very necessary and I am ok, if you only rebase and merge.

domienderle committed:
feat: add tests for doesProjectExistById and doesPaperExistById

domienderle commented:
All right. I have now added the more or less necessary test just for the sake of completeness. Since they were also missing for the method `doesPaperExistById`, I added test for this method, too.

Merseleo reviewed (approved):
There's nothing more to say, very good work.
